# Homework Assignment - Lesson 6: Combining Datasets - Joins

**Student Name:** [Deon Schoeman]

**Due Date:** [10/12/2025]


**Instructions:**
- Complete the following tasks in this R notebook
- Use the exact variable names specified for each task
- Ensure your code runs without errors and produces the expected outputs
- Submit your completed notebook with all code and outputs

---

## Part 1: Data Import and Setup

**Task:** Import the following CSV files and examine their structure.

**Required Variable Names:**
You must use these EXACT variable names:
- `customers` for customers.csv
- `orders` for orders.csv  
- `order_items` for order_items.csv
- `products` for products.csv
- `suppliers` for suppliers.csv

**Expected Outputs:**
- All datasets loaded successfully
- Print dataset dimensions using `nrow()` and `ncol()`
- Show first 3 rows of each dataset using `head(dataset, 3)`

In [2]:
# Your code here:
# 1. Load the tidyverse package
# 2. Set working directory to sample_datasets  
# 3. Import all CSV files using read_csv() with EXACT variable names above
# 4. Print dimensions and head() for each dataset

# Load tidyverse
library(tidyverse)


# Set working directory
setwd("/workspaces/Assignment-3-Data-Transformation-with-dplyr---Part-1/data/")


# Import datasets with exact variable names
customers <- read.csv("customers.csv")
orders <- read.csv("orders.csv")
order_items <- read.csv("order_items.csv")
products <- read.csv("products.csv")
suppliers <- read.csv("suppliers.csv")

# Print dataset information (uncomment after creating variables above):
cat("Dataset Dimensions:\n")
cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
cat("Order Items:", nrow(order_items), "rows x", ncol(order_items), "columns\n")
cat("Products:", nrow(products), "rows x", ncol(products), "columns\n")
cat("Suppliers:", nrow(suppliers), "rows x", ncol(suppliers), "columns\n")

# Show first 3 rows of each dataset (uncomment after loading data):
head(customers, 3)
head(orders, 3)
head(order_items, 3)
head(products, 3)
head(suppliers, 3)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.2     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


Dataset Dimensions:
Customers: 100 rows x 5 columns
Orders: 250 rows x 4 columns
Order Items: 400 rows x 4 columns
Products: 50 rows x 4 columns
Suppliers: 10 rows x 3 columns


CustomerID Name       Email               City        Registration_Date
1 1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2 2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
3 3          Customer 3 customer3@email.com Chicago     2021-04-20

OrderID CustomerID Order_Date Total_Amount
1 1       87         2023-08-30 424.30      
2 2       12         2024-03-24 183.09      
3 3       37         2024-03-19 549.07

OrderID ProductID Quantity Unit_Price
1 213      8        1         43.81    
2 176     18        5        489.16    
3 118      2        5        442.09

ProductID Product_Name Category    Supplier_ID
1 1         Product 1    Home        6          
2 2         Product 2    Electronics 7          
3 3         Product 3    Home        1

Supplier_ID Supplier_Name Country
1 1           Supplier 1    USA    
2 2           Supplier 2    USA    
3 3           Supplier 3    Japan

## Part 2: Basic Joins

**Tasks:**
1. **Inner Join:** Create `customer_orders_inner` by joining customers and orders
2. **Left Join:** Create `customer_orders_left` to show all customers  
3. **Right Join:** Create `customer_orders_right` to show all orders
4. **Full Join:** Create `customer_orders_full` for complete data view

Use the exact variable names specified above for autograding. Analyze row counts and explain results.

In [3]:
# Your code here:
# Create an inner join between customers and orders
customer_orders_inner <- inner_join(customers, orders, by = "CustomerID")
print("Inner Join (matching CustomerIDs only): First 10 rows")
head(customer_orders_inner, 10)
print("Original Customers Data: First 5 rows")
head(customers, 5)
print("Original Orders Data: First 5 rows")
head(orders, 5)

# Analyze the results and compare row counts
cat("Dataset Dimensions:\n")
cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
cat("Customer Orders Inner Join:", nrow(customer_orders_inner), "rows x", ncol(customer_orders_inner), "columns\n")
cat("Inner join data invalid CustomerIDs count:", sum(is.na(customer_orders_inner$CustomerID)), "\n")
cat("Inner join data invalid Name count:", sum(is.na(customer_orders_inner$Name)), "\n")
cat("Inner join data invalid Email count:", sum(is.na(customer_orders_inner$Email)), "\n")
cat("Inner join data invalid City information count:", sum(is.na(customer_orders_inner$City)), "\n")
cat("Inner join data invalid Registration Date count:", sum(is.na(customer_orders_inner$Registration_Date)), "\n")
cat("Inner join data invalid OrderID count:", sum(is.na(customer_orders_inner$OrderID)), "\n")
cat("Inner join data invalid Order Date count:", sum(is.na(customer_orders_inner$Order_Date)), "\n")


# Student Note: Inner joined data lost 50 rows of data from orders because those CustomerIDs did not have a match in the data sets.
# The columns combined to create 8 columns in the new dataset which added the columns from both datasets.


[1] "Inner Join (matching CustomerIDs only): First 10 rows"


CustomerID Name       Email               City        Registration_Date
1  1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2  1          Customer 1 customer1@email.com Phoenix     2020-10-03       
3  2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
4  2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
5  3          Customer 3 customer3@email.com Chicago     2021-04-20       
6  3          Customer 3 customer3@email.com Chicago     2021-04-20       
7  4          Customer 4 customer4@email.com Houston     2022-06-16       
8  4          Customer 4 customer4@email.com Houston     2022-06-16       
9  5          Customer 5 customer5@email.com Phoenix     2023-08-22       
10 5          Customer 5 customer5@email.com Phoenix     2023-08-22       
   OrderID Order_Date Total_Amount
1   87     2023-03-28  716.18     
2  214     2023-09-12 1343.63     
3  173     2024-02-25  159.98     
4  190     2023-04-19 1503.04     
5   29     2023-03-07  441.06     
6  146     2023-04-10   85.28     
7    7     2023-05-01  482.98     
8   61     2024-01-19  666.43     
9   15     2023-12-27  560.36     
10 150     2023-03-28  278.89

[1] "Original Customers Data: First 5 rows"


CustomerID Name       Email               City        Registration_Date
1 1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2 2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
3 3          Customer 3 customer3@email.com Chicago     2021-04-20       
4 4          Customer 4 customer4@email.com Houston     2022-06-16       
5 5          Customer 5 customer5@email.com Phoenix     2023-08-22

[1] "Original Orders Data: First 5 rows"


OrderID CustomerID Order_Date Total_Amount
1 1        87        2023-08-30  424.30     
2 2        12        2024-03-24  183.09     
3 3        37        2024-03-19  549.07     
4 4        57        2023-05-08 1636.73     
5 5       100        2023-08-17  286.60

Dataset Dimensions:
Customers: 100 rows x 5 columns
Orders: 250 rows x 4 columns
Customer Orders Inner Join: 200 rows x 8 columns
Inner join data invalid CustomerIDs count: 0 
Inner join data invalid Name count: 0 
Inner join data invalid Email count: 0 
Inner join data invalid City information count: 0 
Inner join data invalid Registration Date count: 0 
Inner join data invalid OrderID count: 0 
Inner join data invalid Order Date count: 0 


In [4]:
# Your code here:
# Create a left join to keep all customers
customer_orders_left <- left_join(customers, orders, by = "CustomerID")
print("Left Join (matching CustomerIDs only): First 10 rows")
head(customer_orders_left, 10)
print("Original Customers Data: First 5 rows")
head(customers, 5)
print("Original Orders Data: First 5 rows")
head(orders, 5)
# Count how many customers have no orders

cat("Dataset Dimensions:\n")
cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
cat("Customer Orders Left Join:", nrow(customer_orders_left), "rows x", ncol(customer_orders_left), "columns\n")
cat("Left join data invalid CustomerIDs count:", sum(is.na(customer_orders_left$CustomerID)), "\n")
cat("Left join data invalid Name count:", sum(is.na(customer_orders_left$Name)), "\n")
cat("Left join data invalid Email count:", sum(is.na(customer_orders_left$Email)), "\n")
cat("Left join data invalid City information count:", sum(is.na(customer_orders_left$City)), "\n")
cat("Left join data invalid Registration Date count:", sum(is.na(customer_orders_left$Registration_Date)), "\n")
cat("Left join data invalid OrderID count:", sum(is.na(customer_orders_left$OrderID)), "\n")
cat("Left join data invalid Order Date count:", sum(is.na(customer_orders_left$Order_Date)), "\n")


#Student Note: Left joined data had a very similar result to inner join but left join made sure to keep all the data from customers data set.


[1] "Left Join (matching CustomerIDs only): First 10 rows"


CustomerID Name       Email               City        Registration_Date
1  1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2  1          Customer 1 customer1@email.com Phoenix     2020-10-03       
3  2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
4  2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
5  3          Customer 3 customer3@email.com Chicago     2021-04-20       
6  3          Customer 3 customer3@email.com Chicago     2021-04-20       
7  4          Customer 4 customer4@email.com Houston     2022-06-16       
8  4          Customer 4 customer4@email.com Houston     2022-06-16       
9  5          Customer 5 customer5@email.com Phoenix     2023-08-22       
10 5          Customer 5 customer5@email.com Phoenix     2023-08-22       
   OrderID Order_Date Total_Amount
1   87     2023-03-28  716.18     
2  214     2023-09-12 1343.63     
3  173     2024-02-25  159.98     
4  190     2023-04-19 1503.04     
5   29     2023-03-07  441.06     
6  146     2023-04-10   85.28     
7    7     2023-05-01  482.98     
8   61     2024-01-19  666.43     
9   15     2023-12-27  560.36     
10 150     2023-03-28  278.89

[1] "Original Customers Data: First 5 rows"


CustomerID Name       Email               City        Registration_Date
1 1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2 2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
3 3          Customer 3 customer3@email.com Chicago     2021-04-20       
4 4          Customer 4 customer4@email.com Houston     2022-06-16       
5 5          Customer 5 customer5@email.com Phoenix     2023-08-22

[1] "Original Orders Data: First 5 rows"


OrderID CustomerID Order_Date Total_Amount
1 1        87        2023-08-30  424.30     
2 2        12        2024-03-24  183.09     
3 3        37        2024-03-19  549.07     
4 4        57        2023-05-08 1636.73     
5 5       100        2023-08-17  286.60

Dataset Dimensions:
Customers: 100 rows x 5 columns
Orders: 250 rows x 4 columns
Customer Orders Left Join: 200 rows x 8 columns
Left join data invalid CustomerIDs count: 0 
Left join data invalid Name count: 0 
Left join data invalid Email count: 0 
Left join data invalid City information count: 0 
Left join data invalid Registration Date count: 0 
Left join data invalid OrderID count: 0 
Left join data invalid Order Date count: 0 


In [5]:
# Your code here:
# Create a right join to keep all orders
customer_orders_right <- right_join(customers, orders, by = "CustomerID")
print("Right Join (matching CustomerIDs only): First 10 rows")
head(customer_orders_right, 10)
print("Original Customers Data: First 5 rows")
head(customers, 5)
print("Original Orders Data: First 5 rows")
head(orders, 5)

# Check for orders with invalid customer IDs

cat("Dataset Dimensions:\n")
cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
cat("Customer Orders Right Join:", nrow(customer_orders_right), "rows x", ncol(customer_orders_right), "columns\n")
cat("Right join data invalid CustomerIDs count:", sum(is.na(customer_orders_right$CustomerID)), "\n")
cat("Right join data invalid Name count:", sum(is.na(customer_orders_right$Name)), "\n")
cat("Right join data invalid Email count:", sum(is.na(customer_orders_right$Email)), "\n")
cat("Right join data invalid City information count:", sum(is.na(customer_orders_right$City)), "\n")
cat("Right join data invalid Registration Date count:", sum(is.na(customer_orders_right$Registration_Date)), "\n")
cat("Right join data invalid OrderID count:", sum(is.na(customer_orders_right$OrderID)), "\n")
cat("Right join data invalid Order Date count:", sum(is.na(customer_orders_right$Order_Date)), "\n")

#Student Note: The right joined data kept all the orders 250 rows of data but not all of the customers data matched creating 50 rows of NA values
#in the right joined data set.


[1] "Right Join (matching CustomerIDs only): First 10 rows"


CustomerID Name       Email               City        Registration_Date
1  1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2  1          Customer 1 customer1@email.com Phoenix     2020-10-03       
3  2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
4  2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
5  3          Customer 3 customer3@email.com Chicago     2021-04-20       
6  3          Customer 3 customer3@email.com Chicago     2021-04-20       
7  4          Customer 4 customer4@email.com Houston     2022-06-16       
8  4          Customer 4 customer4@email.com Houston     2022-06-16       
9  5          Customer 5 customer5@email.com Phoenix     2023-08-22       
10 5          Customer 5 customer5@email.com Phoenix     2023-08-22       
   OrderID Order_Date Total_Amount
1   87     2023-03-28  716.18     
2  214     2023-09-12 1343.63     
3  173     2024-02-25  159.98     
4  190     2023-04-19 1503.04     
5   29     2023-03-07  441.06     
6  146     2023-04-10   85.28     
7    7     2023-05-01  482.98     
8   61     2024-01-19  666.43     
9   15     2023-12-27  560.36     
10 150     2023-03-28  278.89

[1] "Original Customers Data: First 5 rows"


CustomerID Name       Email               City        Registration_Date
1 1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2 2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
3 3          Customer 3 customer3@email.com Chicago     2021-04-20       
4 4          Customer 4 customer4@email.com Houston     2022-06-16       
5 5          Customer 5 customer5@email.com Phoenix     2023-08-22

[1] "Original Orders Data: First 5 rows"


OrderID CustomerID Order_Date Total_Amount
1 1        87        2023-08-30  424.30     
2 2        12        2024-03-24  183.09     
3 3        37        2024-03-19  549.07     
4 4        57        2023-05-08 1636.73     
5 5       100        2023-08-17  286.60

Dataset Dimensions:
Customers: 100 rows x 5 columns
Orders: 250 rows x 4 columns
Customer Orders Right Join: 250 rows x 8 columns
Right join data invalid CustomerIDs count: 0 
Right join data invalid Name count: 50 
Right join data invalid Email count: 50 
Right join data invalid City information count: 50 
Right join data invalid Registration Date count: 50 
Right join data invalid OrderID count: 0 
Right join data invalid Order Date count: 0 


In [6]:
<VSCode.Cell id="#VSC-0da36c1e" language="markdown">
## Part 2: Basic Joins

**Required Variable Names:**
You must create these EXACT variable names:

**Part 2.1:** Create `customer_orders` using inner_join()
**Part 2.2:** Create `customer_orders_left` using left_join()  
**Part 2.3:** Create `customer_orders_right` using right_join()
**Part 2.4:** Create `customer_orders_full` using full_join()

**Expected Outputs:**
For each join, print:
- Number of rows in the result using `nrow()`
- Comparison with original dataset sizes
- Analysis of unmatched records

ERROR: Error in parse(text = input): <text>:1:1: unexpected '<'
1: <
    ^


In [7]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
customer_orders <- inner_join(customers, orders, by = "CustomerID")


# Required output for autograding:
cat("Inner Join Results: \n")
cat("Customers:", nrow(customers), "\n")
cat("Orders:", nrow(orders), "\n")
cat("Inner Join Result:", nrow(customer_orders), "\n")

Inner Join Results: 
Customers: 100 
Orders: 250 


Inner Join Result: 200 


In [8]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
customer_orders_left <- left_join(customers, orders, by = "CustomerID")


# Required output for autograding:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results: \n")
cat("Total rows:", nrow(customer_orders_left), "\n")
cat("Customers without orders:", customers_without_orders, "\n")

Left Join Results: 
Total rows: 200 
Customers without orders: 0 


In [9]:
# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
customer_orders_right <- right_join(customers, orders, by = "CustomerID")


# Expected output:
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results: \n")
cat("Total rows:", nrow(customer_orders_right), "\n")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "\n")

Right Join Results: 
Total rows: 250 
Orders with invalid customer IDs: 50 


In [10]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
customer_orders_full <- full_join(customers, orders, by = "CustomerID")


# Expected output:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results: \n")
cat("Total rows:", nrow(customer_orders_full), "\n")
cat("Customers without orders:", customers_only, "\n")
cat("Orders without valid customers:", orders_only, "\n")

Full Join Results: 


Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


## Part 3: Multi-Table Joins

**Tasks:**
1. Create a comprehensive dataset by joining `orders`, `customers`, and `order_items`
2. Extend this dataset by adding `products` and `suppliers` information
3. Create a complete supply chain view showing the full customer-to-supplier relationship

Build your joins step by step and examine the results at each stage.

## Part 3: Multi-Table Joins

**Requirements for Autograding:**
You must create these EXACT variable names in order:

1. `orders_items` - Join orders and order_items  
2. `orders_customers_items` - Add customers to above
3. `complete_order_data` - Add products information
4. `complete_data` - Add suppliers for complete supply chain view

**Expected Outputs:**
Print the number of rows for each step to show the join progression.

In [11]:
# Step 1: REQUIRED variable name: orders_items
orders_items <- inner_join(orders, order_items, by = "OrderID")


# Required output for autograding:
cat("Step 1 - Orders with Items:", nrow(orders_items), "rows\n")

Step 1 - Orders with Items: 400 rows


In [12]:
# Step 2: REQUIRED variable name: orders_customers_items  
orders_customers_items <- inner_join(orders_items, customers, by = "CustomerID")


# Required output for autograding:
cat("Step 2 - Add Customers:", nrow(orders_customers_items), "rows\n")

Step 2 - Add Customers: 310 rows


In [13]:
# Step 3: REQUIRED variable name: complete_order_data
complete_order_data <- inner_join(orders_customers_items, products, by = "ProductID")


# Required output for autograding:
cat("Step 3 - Add Products:", nrow(complete_order_data), "rows\n")

Step 3 - Add Products: 310 rows


In [14]:
# Step 4: REQUIRED variable name: complete_data
complete_data <- inner_join(complete_order_data, suppliers, by = "Supplier_ID")


# Required output for autograding:
cat("Step 4 - Complete Supply Chain:", nrow(complete_data), "rows\n")

Step 4 - Complete Supply Chain: 310 rows


## Part 4: Data Quality Analysis

**Tasks:**
1. Use `anti_join()` to find unmatched records between tables
2. Use `semi_join()` to find matched records
3. Check for duplicate keys and analyze their impact on joins
4. Identify and document any data quality issues

This analysis helps understand data integrity and potential issues in your datasets.

In [ ]:
<VSCode.Cell id="#VSC-d7f91846" language="markdown">
## Part 4: Data Quality Analysis

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customers_no_orders` - Use anti_join() to find customers with no orders
2. `orphaned_orders` - Use anti_join() to find orders without customers  
3. `products_never_ordered` - Use anti_join() to find unordered products
4. `active_customers` - Use semi_join() to find customers with orders

**Expected Outputs:**
Print the count of records for each using `nrow()`.

In [15]:
# REQUIRED variable name: customers_no_orders
customers_no_orders <- anti_join(customers, orders, by = "CustomerID")


# Required output for autograding:
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

Customers who never placed an order: 0 


In [16]:
# REQUIRED variable name: orphaned_orders
orphaned_orders <- anti_join(orders, customers, by = "CustomerID")


# Required output for autograding:
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

Orders without corresponding customers: 50 


In [17]:
# REQUIRED variable name: products_never_ordered
products_never_ordered <- anti_join(order_items, products, by = "ProductID")


# Required output for autograding:
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

Products that were never ordered: 0 


In [18]:
# REQUIRED variable name: active_customers
active_customers <- semi_join(customers, orders, by = "CustomerID")


# Required output for autograding:
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who placed at least one order: 100 


In [19]:
# Data integrity checks - count duplicate keys
# Your code here: Calculate and print duplicate counts for each dataset

# Required outputs for autograding (uncomment after creating the calculations above):
duplicate_customers <- sum(duplicated(customers$CustomerID))
duplicate_orders <- sum(duplicated(orders$OrderID))
duplicate_products <- sum(duplicated(products$ProductID))
duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

cat("Data Quality Summary:\n")
cat("Duplicate customer IDs:", duplicate_customers, "\n")
cat("Duplicate order IDs:", duplicate_orders, "\n")
cat("Duplicate product IDs:", duplicate_products, "\n")
cat("Duplicate supplier IDs:", duplicate_suppliers, "\n")

Data Quality Summary:
Duplicate customer IDs: 0 
Duplicate order IDs: 0 
Duplicate product IDs: 0 
Duplicate supplier IDs: 0 


## Part 5: Business Analysis with Joined Data

**Tasks:**
1. Calculate customer lifetime value metrics (total spent, number of orders, average order value)
2. Analyze product performance (total quantity sold, revenue generated, order frequency)
3. Evaluate supplier performance (total revenue, number of products supplied)
4. Create a regional analysis by customer city

Use your joined datasets to generate meaningful business insights.

In [20]:
<VSCode.Cell id="#VSC-f7364f43" language="markdown">
## Part 5: Business Analysis with Joined Data

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customer_metrics` - Customer analysis with Total_Spent, Order_Count, Avg_Order_Value
2. `product_metrics` - Product analysis with Total_Revenue, Total_Quantity, Order_Frequency  
3. `supplier_metrics` - Supplier analysis with Product_Count by supplier
4. `regional_analysis` - Sales analysis by customer city

**Required Columns:**
- customer_metrics: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
- product_metrics: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
- supplier_metrics: Supplier_ID, Supplier_Name, Product_Count
- regional_analysis: City, Total_Sales, Customer_Count, Avg_Customer_Value

ERROR: Error in parse(text = input): <text>:1:1: unexpected '<'
1: <
    ^


In [21]:
# REQUIRED variable name: customer_metrics
# Your code here: Calculate customer lifetime value metrics
# Required columns: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
customer_metrics <- complete_data %>%
    group_by(CustomerID, Name) %>%
    summarise(
        Total_Spent = sum(Total_Amount),
        Order_Count = n(),
        Avg_Order_Value = sum(Total_Amount)/Order_Count,
        .groups = "drop"
    ) %>% 
    select(CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value) %>%
    arrange(desc(Total_Spent))

# Required output for autograding:
cat("Customer Analysis Summary: \n")
cat("Total customers analyzed:", nrow(customer_metrics), "\n")
cat("Top customer total spent: $", round(max(customer_metrics$Total_Spent, na.rm = TRUE), 2), "\n")
print(head(customer_metrics[order(-customer_metrics$Total_Spent), ], 3))

Customer Analysis Summary: 
Total customers analyzed: 94 
Top customer total spent: $ 12115.12 
# A tibble: 3 × 5
  CustomerID Name        Total_Spent Order_Count Avg_Order_Value
       <int> <chr>             <dbl>       <int>           <dbl>
1         88 Customer 88      12115.           7           1731.
2         63 Customer 63      11653.           6           1942.
3         82 Customer 82       9371.           6           1562.


In [35]:
# REQUIRED variable name: product_metrics  
# Your code here: Analyze product performance
# Required columns: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency

product_metrics <- complete_data %>%
    group_by(ProductID, Product_Name) %>%
    summarise(
        Total_Revenue = Quantity * Unit_Price,
        Total_Quantity = sum(Quantity),
        Order_Frequency = n_distinct(OrderID, ProductID),
        .groups = "drop"
    ) %>% 
    select(ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency) %>%
    arrange(desc(Total_Revenue))

# Required output for autograding:
cat("Product Analysis Summary:\n")
cat("Total products analyzed:", nrow(product_metrics), "\n")
cat("Top product revenue: $", round(max(product_metrics$Total_Revenue, na.rm = TRUE), 2), "\n")
print(head(product_metrics[order(-product_metrics$Total_Revenue), ], 3))

Warning message:
“Returning more (or less) than 1 row per `summarise()` group was deprecated in
dplyr 1.1.0.
ℹ Please use `reframe()` instead.
ℹ When switching from `summarise()` to `reframe()`, remember that `reframe()`
  always returns an ungrouped data frame and adjust accordingly.”


Product Analysis Summary:
Total products analyzed: 310 
Top product revenue: $ 2481.65 
# A tibble: 3 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <int> <chr>                <dbl>          <int>           <int>
1         4 Product 4            2482.             12               5
2        18 Product 18           2476.             21               5
3        18 Product 18           2446.             21               5


In [23]:
# REQUIRED variable name: supplier_metrics
# Your code here: Evaluate supplier performance  
# Required columns: Supplier_ID, Supplier_Name, Product_Count
supplier_metrics <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price),
    Product_Count = n_distinct(ProductID),
    Orders_Count = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  select(Supplier_ID, Supplier_Name, Product_Count, Orders_Count, Total_Revenue)



# Required output for autograding:
cat("Supplier Analysis Summary:\n")
cat("Total suppliers analyzed:", nrow(supplier_metrics), "\n")
cat("Max products per supplier:", max(supplier_metrics$Product_Count, na.rm = TRUE), "\n")
print(supplier_metrics[order(-supplier_metrics$Product_Count), ])

Supplier Analysis Summary:
Total suppliers analyzed: 10 
Max products per supplier: 10 


# A tibble: 10 × 5
   Supplier_ID Supplier_Name Product_Count Orders_Count Total_Revenue
         <int> <chr>                 <int>        <int>         <dbl>
 1           7 Supplier 7               10           48        35416.
 2           4 Supplier 4                6           35        25341.
 3          10 Supplier 10               6           35        31875.
 4           1 Supplier 1                5           31        29706.
 5           3 Supplier 3                5           28        25989.
 6           5 Supplier 5                5           43        43434.
 7           6 Supplier 6                5           27        21003.
 8           2 Supplier 2                3           16        10263.
 9           8 Supplier 8                3           11         6805.
10           9 Supplier 9                2           14        14845.


In [24]:
# REQUIRED variable name: regional_analysis
# Your code here: Create regional analysis by customer city
# Required columns: City, Total_Sales, Customer_Count, Avg_Customer_Value
regional_analysis <- complete_data %>%
    group_by(City) %>%
    summarise(
        Total_Sales = sum(Quantity * Unit_Price),
        Customer_Count = n_distinct(CustomerID),
        Avg_Customer_Value = Total_Sales / Customer_Count,
        .groups = "drop"
    )

# Required output for autograding:
cat("Regional Analysis Summary:\n")
cat("Total cities analyzed:", nrow(regional_analysis), "\n")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "\n")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 5 
Highest city sales: $ 72277.52 


# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             23              3143.
2 Phoenix          49333.             19              2596.
3 Los Angeles      48506.             18              2695.
4 Houston          38143.             18              2119.
5 New York         36417.             16              2276.


In [25]:
# Supplier Performance Analysis
supplier_metrics <- complete_data %>%
  filter(!is.na(Supplier_Name)) %>%
  group_by(Supplier_ID, Supplier_Name, Country) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))

cat("Supplier Performance:\n")
print(supplier_metrics)

Supplier Performance:
# A tibble: 10 × 6
   Supplier_ID Supplier_Name Country     Total_Revenue Products_Supplied
         <int> <chr>         <chr>               <dbl>             <int>
 1           5 Supplier 5    Japan              43434.                 5
 2           7 Supplier 7    Germany            35416.                10
 3          10 Supplier 10   Japan              31875.                 6
 4           1 Supplier 1    USA                29706.                 5
 5           3 Supplier 3    Japan              25989.                 5
 6           4 Supplier 4    South Korea        25341.                 6
 7           6 Supplier 6    Japan              21003.                 5
 8           9 Supplier 9    Japan              14845.                 2
 9           2 Supplier 2    USA                10263.                 3
10           8 Supplier 8    Germany             6805.                 3
# ℹ 1 more variable: Orders_Involved <int>


In [42]:
# Regional Analysis by Customer City
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Sales))

cat("Regional Sales Analysis:\n")
print(regional_analysis)

Regional Sales Analysis:


# A tibble: 5 × 5
  City        Total_Sales Customer_Count Order_Count Avg_Customer_Value
  <chr>             <dbl>          <int>       <int>              <dbl>
1 Chicago          72278.             23          36              3143.
2 Phoenix          49333.             19          32              2596.
3 Los Angeles      48506.             18          32              2695.
4 Houston          38143.             18          30              2119.
5 New York         36417.             16          26              2276.


## Part 6: Complex Business Questions

**Tasks:**
Answer the following business questions using your joined datasets:

1. **Customer Segmentation:** Identify your top 10% of customers by total spending
2. **Product Recommendations:** Which products are frequently bought together?
3. **Supplier Dependency:** Which suppliers are most critical to the business?
4. **Market Expansion:** Which cities have high customer counts but low average order values?

Provide data-driven answers with supporting analysis.

## Part 6: Complex Business Questions

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `top_customers` - Top 10% of customers by spending (use quantile function)
2. `product_combinations` - Products frequently bought together  
3. `critical_suppliers` - Suppliers ranked by importance
4. `market_expansion` - Cities with expansion opportunities

**Expected Outputs:**
Each analysis should include specific metrics and rankings as shown in the required outputs below.

In [33]:
# REQUIRED variable name: top_customers
# Your code here: Identify top 10% of customers by total spending
# Use quantile() function to find the 90th percentile threshold


# Required output for autograding (uncomment after completing your analysis):
top_customers <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)
top_ten_customers <- customer_metrics %>%
    filter(Total_Spent >= top_customers) %>%
    summarise(Count = n())

cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_customers, 2), "\n")
cat("Number of top customers:", nrow(top_ten_customers), "\n")

Customer Segmentation Results:
Top 10% spending threshold: $ 6458.29 
Number of top customers: 1 


In [50]:
# REQUIRED variable name: product_combinations
# Your code here: Find products frequently bought together
# Hint: Group by OrderID, filter orders with multiple items
product_combinations <- complete_data %>%
    group_by(OrderID) %>%
    filter(n_distinct(ProductID) > 1) %>%
    summarise(n = n())
    .group = "drop"

# Required output for autograding (uncomment after completing your analysis):
cat("Product Combination Analysis:\n")
cat("Total product combinations found:", nrow(product_combinations), "\n")
if(nrow(product_combinations) > 0) {
cat("Most frequent combination appears", max(product_combinations$n, na.rm = TRUE), "times\n")
}

Product Combination Analysis:
Total product combinations found: 91 
Most frequent combination appears 5 times


In [57]:
# REQUIRED variable name: critical_suppliers
# Your code here: Analyze which suppliers are most critical
# Consider both revenue contribution and product diversity
critical_suppliers <- complete_data %>%
    filter(!is.na(Supplier_Name)) %>%
    group_by(Supplier_ID, Supplier_Name) %>%
    summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved = n_distinct(OrderID),
    .groups = "drop"
    ) %>%
    arrange(desc(Total_Revenue))


# Required output for autograding (uncomment after completing your analysis):
cat("Critical Suppliers Analysis:\n")
cat("Total suppliers analyzed:", nrow(critical_suppliers), "\n")
print("Top 3 most critical suppliers:")
print(head(critical_suppliers, 3))

Critical Suppliers Analysis:
Total suppliers analyzed: 10 
[1] "Top 3 most critical suppliers:"
# A tibble: 3 × 5
  Supplier_ID Supplier_Name Total_Revenue Products_Supplied Orders_Involved
        <int> <chr>                 <dbl>             <int>           <int>
1           5 Supplier 5           43434.                 5              43
2           7 Supplier 7           35416.                10              48
3          10 Supplier 10          31875.                 6              35


In [55]:
# REQUIRED variable name: market_expansion  
# Your code here: Identify market expansion opportunities
# Focus on cities with multiple customers but varying order values
market_expansion <- complete_data %>%
    group_by(City) %>%
    summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
    ) %>%
    arrange(Avg_Customer_Value)


# Required output for autograding (uncomment after completing your analysis):
cat("Market Expansion Analysis:\n")
cat("Cities evaluated for expansion:", nrow(market_expansion), "\n")
print("Top expansion opportunities:")
print(head(market_expansion, 5))

Market Expansion Analysis:
Cities evaluated for expansion: 5 
[1] "Top expansion opportunities:"
# A tibble: 5 × 5
  City        Total_Sales Customer_Count Order_Count Avg_Customer_Value
  <chr>             <dbl>          <int>       <int>              <dbl>
1 Houston          38143.             18          30              2119.
2 New York         36417.             16          26              2276.
3 Phoenix          49333.             19          32              2596.
4 Los Angeles      48506.             18          32              2695.
5 Chicago          72278.             23          36              3143.


## Part 7: Summary and Insights

**Requirements for Autograding:**
Complete the analysis summary below with specific findings from your work. Your responses will be evaluated for completeness and accuracy.

**Grading Criteria:**
- Specific metrics and numbers from your analysis
- Clear identification of data quality issues  
- Actionable business recommendations
- Technical insights about join performance and appropriateness

## Analysis Summary

### Key Findings from Join Operations:
- **Join Efficiency:** [The inner join of customers and orders had a join efficiency of 80% by rows. The left join of customers and orders had a join efficiency of %80 by rows as well. The right join of customers and orders had a join efficiency of 100% by rows however there were a lot of invalid records in the data set. The same thing occured with the full join as well 100% by rows buy a lot of invalid records. The multi inner join had a join efficiency of 77.5% by rows.]
- **Data Relationships:** [Eventhough the data sets were separated. They all had a common identifier or category that allowed the datasets to be combined like customerIDs. So it makes it especially important when setting up a database to ensure that these identifiers are consistent across tables for future joins.]
- **Multi-table Joins:** [The multi-table inner join combined several tables, which initially increased the data set, but soon was decreased to match records that were able to be matched. This created a more complete data set with less null or missing values.]

### Data Quality Issues Discovered:
- **Orphaned Records:** [When right joining/full join on customers and orders there were about 50 missing records see below for information:
Invalid Name count: 50 
Invalid Email count: 50 
Invalid City information count: 50 
Invalid Registration Date count: 50]
- **Missing Data:** [The right join table had invalid records, the full join also had invalid records. This was due to the way the joins were used to ensure no rows were lost during the join process.]
- **Referential Integrity:** [I did not find any foreign key violations found in the data sets. I am not to sure what this means exactly, since this is the first time im hearing of this term.]

### Business Insights:
- **Top Customer:** [Name: Customer 88 Total spent was 12115 average order was 1731.]
- **Best Product:** [Product 4 and total revenue was 2482.]
- **Regional Performance:** [Chicago had the highest sales from the regional analysis with 72278 in sales. The next top selling city was Phoenix  with 49333 in sales. There is a 23,000 difference between the two cities. I believe this would be a good place to check into what Chicago is doing right and see if it can be replicated in other cities.]
- **Supplier Analysis:** [The top suppliers was supplier 5 even though they supply only 5 products they bring in the most revenue. Then the second highest is supplier 7 who also supplies the most different products. This would be a good place to see why customers prefer these two suppliers products.]

### Strategic Recommendations:
- [1. Check into Chicago sales strategies and see if they can be replicated in other cities to boost sales.
   2. Investigate why supplier 5 and 7 products are preferred by customers and see if other suppliers can match their quality or pricing.
   3. Check into what is needed for the expanding markets in Houston, Phoenix, and New York to succeed in those areas. They have the lowest sales but they may not have the right recourses or marketing strategies in place.]

### Technical Learnings:
- **Inner vs Left Joins:** [I used left joins when I wanted keep all the records from the left like customers to orders join. I preferred using the inner joins when I wanted to ensure that only matching records were kept in the data set. This was especially useful when joining multiple tables to ensure that the final data set was as clean and complete as possible. This helped with doing analysis on the data set.]
- **Multi-table Strategy:** [I wanted a complete data set with as few null or missing values as possible. So I used inner joins to make sure that only matching records were kept.]
- **Performance Considerations:** [I think it is important to consider what analysis type you are doing before choosing a join type. If maintaining all row records is important then left, right, or full joins are better. If data integrity and completeness is more important then inner joins are better. However inner joins can lead to significant data loss if there are many unmatched records, so it's a balance between completeness and integrity.]

In [ ]:
# Optional: Save your analysis results (not required for autograding)
# Your code here: Save key datasets to CSV files if desired


# Final summary output for autograding verification:
cat("\n=== HOMEWORK 6 COMPLETION SUMMARY ===\n")
cat("✓ Data Import: 5 datasets loaded\n")
cat("✓ Basic Joins: 4 join types completed\n") 
cat("✓ Multi-table Joins: 4-step progression completed\n")
cat("✓ Data Quality: Anti-joins and semi-joins performed\n")
cat("✓ Business Analysis: Customer, product, supplier, and regional analysis completed\n")
cat("✓ Complex Questions: Advanced business scenarios analyzed\n")
cat("✓ Summary: Comprehensive insights documented\n")
cat("\nAll required variables created for autograding verification.\n")